File: src/lib/color-utils.ts
export type RGB = [number, number, number];
export type Face = 'U' | 'R' | 'F' | 'D' | 'L' | 'B';

export const FACE_ORDER: Face[] = ['U', 'R', 'F', 'D', 'L', 'B'];

export function averageColor(data: Uint8ClampedArray): RGB {
  let r = 0, g = 0, b = 0;
  const total = data.length / 4;
  for (let i = 0; i < data.length; i += 4) {
    r += data[i];
    g += data[i + 1];
    b += data[i + 2];
  }
  return [Math.round(r / total), Math.round(g / total), Math.round(b / total)];
}

export function rgbDistance(a: RGB, b: RGB): number {
  const dr = a[0] - b[0];
  const dg = a[1] - b[1];
  const db = a[2] - b[2];
  return Math.sqrt(dr * dr + dg * dg + db * db);
}

export function rgbToHsv([r, g, b]: RGB): [number, number, number] {
  const rn = r / 255, gn = g / 255, bn = b / 255;
  const max = Math.max(rn, gn, bn);
  const min = Math.min(rn, gn, bn);
  const d = max - min;
  let h = 0;
  if (d !== 0) {
    if (max === rn) h = ((gn - bn) / d) % 6;
    else if (max === gn) h = (bn - rn) / d + 2;
    else h = (rn - gn) / d + 4;
    h *= 60;
    if (h < 0) h += 360;
  }
  const s = max === 0 ? 0 : d / max;
  const v = max;
  return [h, s, v];
}

/**
 * Classify a sticker color into one of the 6 cube face letters using HSV.
 * Robust to lighting variations: uses value/saturation to detect white,
 * hue ranges for chromatic colors.
 */
export function classifyStickerColor(rgb: RGB): Face {
  const [h, s, v] = rgbToHsv(rgb);
  // White: low saturation, high value
  if (s < 0.25 && v > 0.55) return 'U';
  // Very dark / unlit fallback — treat as closest by hue anyway
  // Yellow: ~40-70
  if (h >= 40 && h < 70) return 'D';
  // Orange vs Red disambiguation (both live in low hues 0-40 / 340-360).
  // Orange: brighter, more yellow-leaning hue (~18-40), high value.
  // Red: deeper, lower hue (<18) or wrap-around (>=340), typically darker / more saturated.
  if (h >= 18 && h < 40) return 'L'; // Orange
  if ((h < 18 || h >= 340)) {
    // Within the red/orange overlap zone (h 10-18), use brightness:
    // bright + less saturated → orange; dark + saturated → red.
    if (h >= 10 && h < 18 && v > 0.75 && s < 0.85) return 'L';
    return 'R'; // Red
  }
  // Green: 70-170
  if (h >= 70 && h < 170) return 'F';
  // Blue: 170-260
  if (h >= 170 && h < 260) return 'B';
  // Magenta-ish → likely red under cool light
  return 'R';
}

export function rotateGrid<T>(cells: T[], times: number): T[] {
  // Rotate 3x3 grid clockwise 'times' times
  const t = ((times % 4) + 4) % 4;
  let grid = [...cells];
  for (let k = 0; k < t; k++) {
    grid = [
      grid[6], grid[3], grid[0],
      grid[7], grid[4], grid[1],
      grid[8], grid[5], grid[2],
    ];
  }
  return grid;
}

export function sample3x3Averages(img: HTMLImageElement): Promise<RGB[]> {
  return new Promise((resolve, reject) => {
    const canvas = document.createElement('canvas');
    const ctx = canvas.getContext('2d');
    if (!ctx) return reject(new Error('Canvas 2D context unavailable'));

    const maxDim = 768;
    let w = img.naturalWidth;
    let h = img.naturalHeight;
    if (w > maxDim || h > maxDim) {
      if (w > h) {
        h = Math.round((h * maxDim) / w);
        w = maxDim;
      } else {
        w = Math.round((w * maxDim) / h);
        h = maxDim;
      }
    }
    canvas.width = w;
    canvas.height = h;
    ctx.drawImage(img, 0, 0, w, h);

    const cellW = Math.floor(w / 3);
    const cellH = Math.floor(h / 3);
    const padX = Math.floor(cellW * 0.15);
    const padY = Math.floor(cellH * 0.15);

    const rgbs: RGB[] = [];
    for (let row = 0; row < 3; row++) {
      for (let col = 0; col < 3; col++) {
        const x = col * cellW + padX;
        const y = row * cellH + padY;
        const sw = Math.max(1, cellW - padX * 2);
        const sh = Math.max(1, cellH - padY * 2);
        const imgData = ctx.getImageData(x, y, sw, sh);
        rgbs.push(averageColor(imgData.data));
      }
    }
    resolve(rgbs);
  });
}

File: android/app/src/main/java/com/cubesolverai/app/colorutils/ColorUtils.kt
package com.cubesolverai.app.colorutils

import android.graphics.Bitmap
import com.cubesolverai.app.model.CubeFace
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * 1:1 Kotlin port of the web app's src/lib/color-utils.ts. Same downscale/padding constants
 * and HSV thresholds, so a photo classifies identically on both platforms. Do not "improve"
 * the thresholds here without also updating the web app - they were tuned empirically there.
 */
data class RGB(val r: Int, val g: Int, val b: Int)

fun rgbToHsv(rgb: RGB): Triple<Float, Float, Float> {
    val rn = rgb.r / 255f
    val gn = rgb.g / 255f
    val bn = rgb.b / 255f
    val maxV = max(rn, max(gn, bn))
    val minV = min(rn, min(gn, bn))
    val d = maxV - minV
    var h = 0f
    if (d != 0f) {
        h = when (maxV) {
            rn -> ((gn - bn) / d) % 6f
            gn -> (bn - rn) / d + 2f
            else -> (rn - gn) / d + 4f
        }
        h *= 60f
        if (h < 0f) h += 360f
    }
    val s = if (maxV == 0f) 0f else d / maxV
    val v = maxV
    return Triple(h, s, v)
}

/** Classify a sticker color into one of the 6 cube face letters using HSV. */
fun classifyStickerColor(rgb: RGB): CubeFace {
    val (h, s, v) = rgbToHsv(rgb)
    // White: low saturation, high value
    if (s < 0.25f && v > 0.55f) return CubeFace.U
    // Yellow: ~40-70
    if (h >= 40f && h < 70f) return CubeFace.D
    // Orange: brighter, more yellow-leaning hue (~18-40)
    if (h >= 18f && h < 40f) return CubeFace.L
    if (h < 18f || h >= 340f) {
        // Orange/red overlap zone (h 10-18): bright + less saturated -> orange; else red.
        if (h >= 10f && h < 18f && v > 0.75f && s < 0.85f) return CubeFace.L
        return CubeFace.R
    }
    // Green: 70-170
    if (h >= 70f && h < 170f) return CubeFace.F
    // Blue: 170-260
    if (h >= 170f && h < 260f) return CubeFace.B
    // Magenta-ish -> likely red under cool light
    return CubeFace.R
}

fun <T> rotateGrid(cells: List<T>, times: Int): List<T> {
    val t = ((times % 4) + 4) % 4
    var grid = cells
    repeat(t) {
        grid = listOf(
            grid[6], grid[3], grid[0],
            grid[7], grid[4], grid[1],
            grid[8], grid[5], grid[2],
        )
    }
    return grid
}

/** Downscales to max 768px, splits into a 3x3 grid, averages RGB over the center 70% of each cell. */
fun sample3x3Averages(bitmap: Bitmap): List<RGB> {
    val maxDim = 768
    var w = bitmap.width
    var h = bitmap.height
    if (w > maxDim || h > maxDim) {
        if (w > h) {
            h = ((h.toLong() * maxDim) / w).toInt()
            w = maxDim
        } else {
            w = ((w.toLong() * maxDim) / h).toInt()
            h = maxDim
        }
    }
    val scaled = if (w == bitmap.width && h == bitmap.height) {
        bitmap
    } else {
        Bitmap.createScaledBitmap(bitmap, w, h, true)
    }

    val cellW = w / 3
    val cellH = h / 3
    val padX = (cellW * 0.15).toInt()
    val padY = (cellH * 0.15).toInt()

    val result = mutableListOf<RGB>()
    for (row in 0..2) {
        for (col in 0..2) {
            val x = col * cellW + padX
            val y = row * cellH + padY
            val sw = max(1, cellW - padX * 2)
            val sh = max(1, cellH - padY * 2)
            result.add(averageColor(scaled, x, y, sw, sh))
        }
    }
    if (scaled !== bitmap) scaled.recycle()
    return result
}

private fun averageColor(bitmap: Bitmap, x: Int, y: Int, w: Int, h: Int): RGB {
    val clampedW = min(w, bitmap.width - x)
    val clampedH = min(h, bitmap.height - y)
    val pixels = IntArray(clampedW * clampedH)
    bitmap.getPixels(pixels, 0, clampedW, x, y, clampedW, clampedH)
    var rSum = 0L
    var gSum = 0L
    var bSum = 0L
    for (p in pixels) {
        rSum += (p shr 16) and 0xFF
        gSum += (p shr 8) and 0xFF
        bSum += p and 0xFF
    }
    val total = pixels.size.coerceAtLeast(1)
    return RGB(
        (rSum.toDouble() / total).roundToInt(),
        (gSum.toDouble() / total).roundToInt(),
        (bSum.toDouble() / total).roundToInt(),
    )
}
